crates/tabby-scheduler/src/index.rs

```diff
@@ -71,9 +79,101 @@ pub fn index_repositories(_config: &[RepositoryConfig]) {
         .expect("Failed to wait for merging threads");
 }
 
-fn is_valid_file(file: &SourceFile) -> bool {
-    file.max_line_length <= MAX_LINE_LENGTH_THRESHOLD
-        && file.avg_line_length <= AVG_LINE_LENGTH_THRESHOLD
+fn index_repository_from_scratch(
+    repository: &RepositoryConfig,
+    writer: &IndexWriter,
+    code: &CodeSearchSchema,
+    intelligence: &CodeIntelligence,
+    cache: &mut CacheStore,
+) {
+    let mut pb = std::io::stdout().is_terminal().then(|| {
+        let total_file_size: usize = Walk::new(repository.dir())
+            .filter_map(|f| f.ok())
+            .map(|f| f.path().to_owned())
+            .filter(|f| {
+                f.extension()
+                    .is_some_and(|ext| get_language_by_ext(ext).is_some())
+            })
+            .map(|f| read_file_size(&f))
+            .sum();
+        tqdm(total_file_size)
+    });
+
+    for file in Walk::new(repository.dir()) {
+        let file = file.expect("Failed to read file listing");
+        let Some(source_file) = cache.get_source_file(repository, file.path()) else {
+            continue;
+        };
+        if !is_valid_file(&source_file) {
+            continue;
+        }
+        add_indexed_source_file(writer, repository, &source_file, code, intelligence);
+        pb.as_mut().map(|pb| {
+            pb.update(source_file.read_file_size())
+                .expect("Failed to update progress bar")
+        });
+    }
+    cache.set_last_index_commit(
+        repository,
+        Some(
+            get_current_commit_hash(&repository.dir()).expect("Failed to read current commit hash"),
+        ),
+    );
+}
+
+fn read_file_size(path: &Path) -> usize {
+    std::fs::metadata(path)
+        .map(|meta| meta.len())
+        .unwrap_or_default() as usize
+}
+
+fn is_valid_file(source_file: &SourceFile) -> bool {
+    source_file.max_line_length <= MAX_LINE_LENGTH_THRESHOLD
+        && source_file.avg_line_length <= AVG_LINE_LENGTH_THRESHOLD
+}
+
+pub fn delete_indexed_source_file(
```

Comment:
```suggestion
pub fn delete_document_from_index(
```

Reply:
Using `delete_document` for consistent naming with your other suggestion